src/protocol/sort/generate_permutation_opt.rs

```diff
@@ -109,39 +114,159 @@ where
+/// In the end, following is returned
+///    i. n-1th composition: This is the permutation which sorts the inputs
+///   ii. Validator which can be used to validate the leftover items in the accumulator
+///
+/// # Panics
+/// If sort keys dont have num of bits same as `num_bits`
+/// # Errors
+pub async fn malicious_generate_permutation_opt<'a, F>(
+    sh_ctx: SemiHonestContext<'a, F>,
+    sort_keys: &[Vec<Replicated<F>>],
+    num_bits: u32,
+    num_multi_bits: u32,
+) -> Result<(MaliciousValidator<'a, F>, Vec<MaliciousReplicated<F>>), Error>
+where
+    F: Field,
+{
+    let mut malicious_validator = MaliciousValidator::new(sh_ctx.narrow(&MaliciousUpgradeContext));
+    let mut m_ctx = malicious_validator.context();
+    let m_ctx_0 = m_ctx.narrow(&Sort(0));
+    assert_eq!(sort_keys.len(), num_bits as usize);
+
+    let last_bit_num = std::cmp::min(num_multi_bits, num_bits);
+
+    let upgraded_sort_keys = try_join_all((0..last_bit_num).zip(repeat(m_ctx.clone())).map(
+        |(i, m_ctx)| async move {
+            m_ctx
+                .upgrade_vector(&MaliciousUpgradeInput(i), sort_keys[i as usize].clone())
+                .await
+        },
+    ))
+    .await?;
+    let lsb_permutation =
+        multi_bit_permutation(m_ctx_0.narrow(&BitPermutationStep), &upgraded_sort_keys).await?;
+    let input_len = u32::try_from(sort_keys[0].len()).unwrap(); // safe, we don't sort more that 1B rows
+
+    let mut composed_less_significant_bits_permutation = lsb_permutation;
+    for bit_num in (num_multi_bits..num_bits).step_by(num_multi_bits.try_into().unwrap()) {
+        let mut m_ctx_bit = m_ctx.narrow(&Sort(bit_num));
+        let revealed_and_random_permutations = malicious_shuffle_and_reveal_permutation(
+            m_ctx_bit.narrow(&ShuffleRevealPermutation),
+            input_len,
+            composed_less_significant_bits_permutation,
+            malicious_validator,
+        )
+        .await?;
+
+        malicious_validator = MaliciousValidator::new(sh_ctx.narrow(&Sort(bit_num)));
+        m_ctx_bit = malicious_validator.context();
+
+        let last_bit_num = std::cmp::min(bit_num + num_multi_bits, num_bits);
+        let upgraded_sort_keys =
+            &try_join_all((bit_num..last_bit_num).zip(repeat(m_ctx_bit.clone())).map(
+                |(i, m_ctx_bit)| async move {
+                    m_ctx_bit
+                        .upgrade_vector(&MaliciousUpgradeInput(i), sort_keys[i as usize].clone())
+                        .await
+                },
+            ))
+            .await?;
+
+        let (randoms_for_shuffle0, randoms_for_shuffle1, revealed) = (
+            revealed_and_random_permutations
+                .randoms_for_shuffle
+                .0
+                .as_slice(),
+            revealed_and_random_permutations
+                .randoms_for_shuffle
+                .1
+                .as_slice(),
+            revealed_and_random_permutations.revealed.as_slice(),
+        );
```

Comment:
Is it possible to do this destructuring when `revealed_and_random_permutations` is originally created back up on line 184?